File: old/dekaf-core/src/main/java/org/jetbrains/dekaf/sql/TextPointer.java
package org.jetbrains.dekaf.sql;

/**
 * @author Leonid Bushuev from JetBrains
 */
final class TextPointer {

  /**
   * Offset from the beginning of the text.
   */
  final int offset;

  /**
   * Row number, starts with 1.
   */
  final int row;

  /**
   * Number of positions inside the row, starts with 1.
   */
  final int pos;


  TextPointer(int offset, int row, int pos) {
    this.offset = offset;
    this.row = row;
    this.pos = pos;
  }


  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    TextPointer pointer = (TextPointer) o;

    return offset == pointer.offset && pos == pointer.pos && row == pointer.row;
  }


  @Override
  public int hashCode() {
    return offset;
  }
}

File: dekaf-main/src/text/TextPointer.kt
package org.jetbrains.dekaf.text

/**
 * Text Pointer - a pointer to a place in a text.
 * @author Leonid Bushuev
 */
data class TextPointer 
(
        /**
         * Offset from the beginning of the text.
         */
        val offset: Int,

        /**
         * Row number, starts with 1.
         */
        val row: Int,

        /**
         * Number of positions inside the row, starts with 1.
         */
        val pos: Int
)
{

    override fun hashCode(): Int {
        return offset
    }

    override fun toString(): String {
        return "$row:$pos($offset)"
    }
    
}
